### PR TITLE
Add method-level sources to CartesianTest

### PR DESCRIPTION
**Pull request by Michael1993**
Forget the checklist and the commit message. I'll add them later when this is actually ready for merge. For now this is just an idea I need feedback on.

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

### CODE REVIEW

**Comment by Bukama** [comment]
At our team meeting we talked about some examples when to use which annotation and if they are combinable.

@nipafx and @beatngu13 also wanted to have a deeper look at the code, if I remember correctly.

**Comment by nipafx** [comment]
Can you describe the idea in a few words? :)

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianEnumArgumentsProvider.java
Comment (nipafx) [reply src/main/java/org/junitpioneer/jupiter/cartesian/CartesianEnumArgumentsProvider.java:30]: Oh, I proposed to make `CartesianArgumentsProvider` package-visible. Then the comment could go into `package-info.java`. Having read through `CartesianMethodArgumentsProvider` and `CartesianParameterArgumentsProvider`, I don't think we should stress the connection to `ArgumentsProvider` too much or even encourage people to implement both.

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTestExtension.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTestExtension.java:91]: I think the formatter is fine with the code as is. 😉

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTestExtension.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTestExtension.java:122]: Shouldn't this then be an `IllegalStateException`? I understand `PreconditionViolationException` to mean that the user did something wrong.

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTestExtension.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTestExtension.java:142]: As above.

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTest.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTest.java:316]: Should we call this `MethodSource` in line with Jupiter or is there a good reason to have a different name (e.g. to avoid collisions)?

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianMethodArgumentsProvider.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/cartesian/CartesianMethodArgumentsProvider.java:43]: This class and its methods need Javadoc.

File: docs/cartesian-product.adoc
Comment (nipafx) [inline docs/cartesian-product.adoc:314]: It should be made clearer that `add` needs to be called once per parameter with all arguments for that parameter.

File: docs/cartesian-product.adoc
Comment (nipafx) [inline docs/cartesian-product.adoc:310]: Is this line still up to date? I can't see `name` on `Factory`.

File: docs/cartesian-product.adoc
Comment (nipafx) [inline docs/cartesian-product.adoc:387]: Personally, I'd prefer "requirements", but I'll leave that to you.

File: docs/cartesian-product.adoc
Comment (nipafx) [inline docs/cartesian-product.adoc:393]: What if we could inject parameters into factory methods, though?! 🤡

File: docs/cartesian-product.adoc
Comment (nipafx) [inline docs/cartesian-product.adoc:453]: There's a later comment in this review where I give an opinion on when and how to mention `ArgumentsProvider`. As to how that applies here: I think this explanation here is good and should stay as is, but all other mentions of `ArgumentsProvider` in this documentation should be removed.

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTest.java
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTest.java:316]: No, we should not. For several reasons:
 - In `@CartesianProductTest` this was a property called `factory`.
 - It avoids collision, but more importantly it avoids confusion. This annotation is only for `@CartesianTest`.
 - We decided to use short names for simplicity. Even though this annotation doesn't go on a parameter, I'd still keep it short.
 - I don't want to change it. 😭

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianEnumArgumentsProvider.java
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/cartesian/CartesianEnumArgumentsProvider.java:30]: Tbh. there is actually no real connection to `ArgumentsProvider`...

File: docs/cartesian-product.adoc
Comment (Michael1993) [reply docs/cartesian-product.adoc:393]: Ha-ha, very funny.

File: docs/cartesian-product.adoc
Comment (Michael1993) [reply docs/cartesian-product.adoc:453]: I think this is the _only_ explanation that's needed.

File: docs/cartesian-product.adoc
Comment (Michael1993) [reply docs/cartesian-product.adoc:314]: Updated it a bit.

Comment (nipafx) [reply docs/cartesian-product.adoc:314]: Me too. 😁

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianEnumArgumentsProvider.java
Comment (nipafx) [reply src/main/java/org/junitpioneer/jupiter/cartesian/CartesianEnumArgumentsProvider.java]: Good thing it's gone now. :)

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianEnumArgumentsProvider.java
Comment (nipafx) [reply src/main/java/org/junitpioneer/jupiter/cartesian/CartesianEnumArgumentsProvider.java:33]: > the exception message is from the future

Lol.

Since I'm already editing, I'll replace the message.

File: src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTest.java
Comment (nipafx) [reply src/main/java/org/junitpioneer/jupiter/cartesian/CartesianTest.java:316]: I have to think about it. If I feel we need to change it, I can do the work. 😉

**Comment by nipafx** [comment]
Looks great! Things left to do:

* make sure correct exceptions are thrown (I see a strong preference of `PreconditionViolationException`)
* further eradicate mentions of `ArgumentsProvider`
* consider renaming/refactoring `Sets` for better ease of use

I can tackle those on stream in an hour.

File: docs/cartesian-product.adoc
Comment (filiphr) [inline docs/cartesian-product.adoc:384]: I don't understand why you don't piggy back on what is already available there and use `MethodSource`. I can write this like:

```
static Stream<Arguments> provideArguments() {
    return Stream.of(
        Arguments.of("Mercury", "Earth", "Venus"),
        Arguments.of(1, 12, 144),
    )
}
```

You will benefit of a lot of simplifications:

1. No need to do magic with reflection and you piggy back on the logic done by JUnit Jupiter (not sure if you can get an instance of the provider though)
2. You get nice auto completion and methods marked as used in IntelliJ

Comment (nipafx) [reply docs/cartesian-product.adoc:384]: First and foremost, because `Arguments` is not the right abstraction. [Its Javadoc](https://junit.org/junit5/docs/current/api/org.junit.jupiter.params/org/junit/jupiter/params/provider/Arguments.html) says:

> `Arguments` is an abstraction that provides access to an array of objects to be used for invoking a `@ParameterizedTest` method. 

What we need, though, is one "array of objects" per parameter (not per execution) and we're not invoking anything with them (they get folded into other "arrays", which then dictate the number of executions), let alone a `@ParameterizedTest`.

Re 1.: Unless we're reflectively accessing `MethodArgumentsProvider`, we still have to write our own reflective code to get these arguments. Even if we copy its source code, we'd still have to maintain it ourselves.

Re 2.: True, but I don't think that's worth the trade off.

**Comment by nipafx** [comment]
All that's needed is a commit message, right?

**Comment by Michael1993** [comment]
> All that's needed is a commit message, right?

Yes. Well, also approvals from maintainers. 😄

**Review by nipafx — APPROVED** [review]

**Pull request merged**
